### ISSUE TITLE
[FEATURE] Complete Cluster Backup

### ISSUE THREAD

**Issue by rebeccazzzz** [harvester/harvester#3263]
**Is your feature request related to a problem? Please describe.**
Users who lose the entire Harvester cluster will find it difficult to restore service if they do not have the VM metadata to pair with the volume backups.

**Describe the solution you'd like**
As a Harvester cluster admin, I need to be able to restore service to my cluster retaining all the original configuration as quickly as possible in the event of: 
- etcd database corruption
- an administrative mistake that erases key configuration data such as deleting system components
- a malicious user gaining access to the system and purposefully wiping out the cluster
- catastrophic loss of control plane hardware 

"Complete cluster backup" should include:
- The etcd database for the RKE2/Rancher cluster supporting. (Suggest we could use the existing Rancher backup-restore operator)
- The OS settings that would be needed to restore the OS to the same state at the time of backup.
  - OS setting should include the network configuration such as NIC assignments, bonding, etc. 

**Additional context**
Out of scope for this ticket should be the VM workloads themselves. VM workloads should be backed up through the separate VM backup workflow in Harvester (https://docs.harvesterhci.io/v1.1/vm/backup-restore). This also means guest Kubernetes clusters deployed by Rancher would need to be backed up as VMs and in Rancher itself (or redeployed by the user after restoring the Harvester cluster).

**Comment by Martin-Weiss** [comment]
@rebeccazzzz 
- For the installation of the cluster I use the automated way with control files. So in case we lost the cluster we can re-deploy. 
- For the harvester config I use yaml or terraform where it is not possible with the installation control files. 
- For the image deployment I also use terraform.

When I backup the VMs in Harvester to S3 or NFS, now - I can use the UI to restore the VMs directly and do not need to know VM metadata/volume backups etc.

What exactly are you missing, here?

**Comment by tlehman** [comment]
I am also wondering too, this issue seems underspecified

**Comment by rebeccazzzz** [comment]
The description above has been updated with additional information @tlehman

**Comment by tlehman** [comment]
Would a full byte-by-byte backup of each node in the cluster be acceptable? That is, I can `dd` the entire drive at some point and then that image can be restored. That would give the VM workload backups for free, but the cost would be a very large backup file.

For a small backup file, we can [snapshot the etcd](https://etcd.io/docs/v3.3/op-guide/recovery/#snapshotting-the-keyspace) and then get all the YAML configs for the cluster and the OS config files for the nodes. That would be considerably smaller, the remaining question is how to restore it.

For the small backup, there are two choices:
1. modify the installer to point at the backup files, then have the installer unpack all the configs as it goes
2. leave the installer alone and then create a backup-and-restore controller for Harvester.
   (the reason I put this as a question is that applying the yaml configs after the cluster is set up is straightforward, but changing OS configs needs to be done with care, and having a process on the computer that can change arbitrary OS configs should be done lightly). Doing the restoration in the installer seems more appropriate but I am still open to arguments otherwise

I'll ask around and see which of these is a better option

**Comment by wjimenez5271** [comment]
I don't think a `dd` approach is going to be what we need, because if the user is already backing up their VMs they would have redundant backups of their VMs. 

For the other method and the subsequent "restore" question, I vote for option 1 with the installer unpacking the backup files. Probably should source some other inputs people like Alejandro B, @Martin-Weiss, @janeczku

**Comment by Martin-Weiss** [comment]
Just a quick brainstorm:

For backing up the "whole" harvester cluster I guess we need the automated installation - BUT - anything that a user might have changed in the UI after installation will be missing... so we might need a "export current config" to a backup target in a scheduled way so that we can re-install with the "latest config status" of the harvester cluster. (backup target NFS / S3)

I also believe that in some situations we might need the harvester internal RKE2 ETCD database snapshot / backup -> also this one could be stored in a scheduled way on a central S3 or NFS target.

Last but not least - we need a scheduled way for backing up virtual machines and longhorn volumes.

---

Once we have "everything backed up" we need to define the disaster recovery scenarios and restore procedures (lost node, lost cluster, lost config setting)

-> importing a backup of a config would be great (would allow a rollback of known-good settings - settings page)
-> restoring etcd snapshot (already part of RKE2 / Rancher) for corrupt etcd database or deleted data
-> VM restore / Longhorn volume restore is possible, already - but the integration with CSI in downstream clusters and details about how to re-link to existing VMs might need some docs..

**Comment by ibrokethecloud** [comment]
I am more in favor of supporting full cluster configuration using fleet (gitops)

This should make it easier to recreate the entire cluster state from a git repo.

Restoring from etcd snapshots may have issues with state of longhorn volumes, and eventually data consistency in those volumes.

**Comment by ibrokethecloud** [comment]
Please add more, but at a bare minimum we should have gitops based support for following objects:

* Cluster network definitions
* VM network definitions
* VirtualMachineImages
* Backup targets
* Cloud Config
* SSH Keys
* Additional Storage Classes
* Addon configuration
* Settings
* Logging configuration
* Monitoring configuration

In addition, we need a means to expose OS configuration by wrapping `yip` into a k8s controller, or other means to reconcile OS level changes with a gitrepo.

**Comment by w13915984028** [comment]
connect this feature with following
epic: https://github.com/harvester/harvester/issues/3261
sub-story: https://github.com/harvester/harvester/issues/4450

**Comment by ibrokethecloud** [comment]
depends on: https://github.com/longhorn/longhorn/issues/5085

**Comment by ibrokethecloud** [comment]
currently system restore is broken due to a minor issue with ordering of restored objects. As a result longhorn system restore never completes and  is being tracked: https://github.com/longhorn/longhorn/issues/8601

**Comment by bk201** [comment]
Related: https://github.com/longhorn/longhorn/issues/8483

**Comment by w13915984028** [comment]
Some questions/thoughts:

(1) LH [system backup and restore](https://longhorn.io/docs/1.7.2/advanced-resources/system-backup-restore/) looks to be more focused on LH itself, can this feature be freely expanded to the whole cluster with anything?

(2) If using gitops to backup all CRD objects, how to make sure the sequences/dependencies to restore them, e.g.:
- certs
- the disks added by node-manager and provisioned to LH
- cluster network & vm network
- storage network
- LH volumes
- those dynamic objects created by controllers like VM IP allocation record; the machines mapping the nodes
- the `owner references with ID` on many objects

the cluster is built step by step, the restore looks to be similar:

(manually) install/restore a blank cluster first, then deploy all the workloads per defined sequences

(3) The `rke2 etcd back & restore` based solution

the pre-condition is the node & disk & network keeps same;  restore a cluster either on the original nodes or new nodes

suppose the `harvester-installer` is responsible for recovering all the `oem` path configure files to help the OS restoring first, then restore the rke2 etcd db

when (2)/(3) restores the cluster, the final step(s) is to restore the VMs based on current VM back&restore solution

**Comment by belgaied2** [comment]
If I can give some user feedback here, this is what the user story should look like:

> As a platform operator, I would like to recover my **workloads** on **a new environment** if I encounter a total failure.

This would mean multiple things:
- Restore Hardware might be completely different from initial hardware
- Potentially, IP scheme for nodes and VM network **might** also be different, but this is less important than the above
- Recovering Longhorn disk structure and replicas locations etc. is not relevant, only the data counts.
- As mentioned in [Martin's comment](https://github.com/harvester/harvester/issues/3263#issuecomment-1343939364), everything already exists in Harvester. However, my opinion is that it is not well documented enough, and it is up to the user to tie multiple components together.
- Another question would be: should this also trigger VM Backups or only rely on existing one ?
   - If Cluster Backup does trigger a backup of all VMs in all namespaces, then we might encounter backup storage limitations if the user did not plan for it. It might also make the backup last multiple **days**!
   - If Cluster Backup does NOT trigger a backup of all VMs, then we don't really have a hands-off solution, but that might be okay if it documented correctly.

**Suggestion:**
Automate recovery of the items mentioned by @ibrokethecloud in [this](https://github.com/harvester/harvester/issues/3263#issuecomment-1439301146) comment and others such as Load Balancer and IPPool definitions and other CRDs such as ippools of DHCP Controller and any other potential CRs from "official" Addons.

Ideally the user would add a setting `cluster-backup-url` to the UI and/or to the install configuration of the first server, and the cluster will handle everything.
